registry/config.go

```diff
@@ -90,10 +107,104 @@ func newServiceConfig(options ServiceOptions) (*serviceConfig, error) {
 	if err := config.LoadInsecureRegistries(options.InsecureRegistries); err != nil {
 		return nil, err
 	}
+	if err := config.LoadRegistries(options.Registries); err != nil {
+		return nil, err
+	}
 
 	return config, nil
 }
 
+// checkRegistries makes sure that no mirror serves more than one registry and
+// that no host is used as a registry and as a mirror simultaneously.  Notice
+// that different registry prefixes can share a mirror as long as they point to
+// the same registry.  It also warns if the URI schemes of a given registry and
+// one of its mirrors differ.
+func (config *serviceConfig) checkRegistries() error {
+	inUse := make(map[string]string) // key: host, value: user
+
+	// make sure that each mirror serves only one registry
+	for _, reg := range config.Registries {
+		for _, mirror := range reg.Mirrors {
+			if used, conflict := inUse[mirror.URL.Host()]; conflict {
+				if used != reg.URL.Host() {
+					return fmt.Errorf("mirror '%s' can only serve one registry host", mirror.URL.Host())
+				}
+			}
+			// docker.io etc. is reserved
+			if mirror.URL.IsOfficial() {
+				return fmt.Errorf("mirror '%s' cannot be used (reserved host)", mirror.URL.Host())
```

Comment:
That is correct, but I am not sure if this would be without danger as there is a probability of image name conflicts etc. when two registries are mixed. Having a clear distinction between what a mirror and what registry is avoids any potential problems and conflicts in the first place.

In the upper example, I would argue that the relation between `docker.io` and `my-registry.internal` is inverted and that the latter should serve as a mirror for `docker.io`. To the best of my knowledge, `docker/distribution` allows a mirror to only serve as a pull-through cache.

Reply:
>  I am not sure if this would be without danger as there is a probability of image name conflicts etc. when two registries are mixed

Yes, I know this has historically been quite some discussions around that. However (as I see it), this is already a problem with allowing a local registry acting a as (pull-through) mirror for Docker Hub; I can push any random image to my local mirror, and have it mask the actual image on Docker Hub (example below).

If we _could_ drop this restriction, it makes the mirror configuration un-opinionated, removes the special handling of the "official registry", and possibly could serve some use-cases, for example to use a local registry as a backup for Docker Hub (in case of an outage of Docker Hub, the local registry will be used)


### Example "name conflict" with local mirror

Start a local registry, and configure it as mirror for Docker Hub;

```bash
docker run -d -p 5000:5000 registry:2 

echo '{"debug":true,"registry-mirrors":["http://localhost:5000"]}' > /etc/docker/daemon.json
killall -HUP dockerd
```

Tag, and push busybox as "ubuntu"

```bash
docker image tag busybox localhost:5000/library/ubuntu:14.04  
docker image push localhost:5000/library/ubuntu:14.04 
```

Now (if I don't have `DOCKER_CONTENT_TRUST` enabled), the image will be pulled from the mirror, and presented as `ubuntu:14.04`:


```bash
docker image pull ubuntu:14.04
```

```bash
docker image ls

REPOSITORY                             TAG                                       IMAGE ID            CREATED             SIZE
busybox                                latest                                    f9b6f7f7b9d3        37 hours ago        1.14MB
ubuntu                                 14.04                                     f9b6f7f7b9d3        37 hours ago        1.14MB
localhost:5000/library/ubuntu          14.04                                     f9b6f7f7b9d3        37 hours ago        1.14MB
```

Reply:
If I may summarize your findings: we should not make any restrictions on how a given host is used (a) to give the users more flexibility and (b) we can already work around that (as shown in the example above). It makes a lot of sense to me, thanks for explaining that in such great detail.

Reply:
Thanks! Yes, that's how I look at it (but of course, it can be further discussed during progress of this PR - would like other maintainers to double check that we're not overlooking something)